Add Statevector.from_circuit to mirror Operator.from_circuit

debasmita2102 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds a `from_circuit` class method to `Statevector` to mirror the existing `Operator.from_circuit` functionality. This addresses issue #13621 where users experienced inconsistent statevector results when transpiling circuits at different optimization levels.

**Fixes #13621**

### Details and comments

The new `Statevector.from_circuit(circuit, input_state=None)` method:
- Creates a statevector by evolving an input state through a quantum circuit
- Properly handles circuit layouts to ensure consistent qubit ordering 
- Accepts both `QuantumCircuit` and `Instruction` objects
- Defaults to the zero state `|0...0⟩` if no input state is provided
- Leverages the existing `Operator.from_circuit` implementation for consistency

**Problem solved:**
Previously, `Statevector(transpiled_circuit)` would give different results depending on the transpilation optimization level due to virtual qubit remapping. 

**Example demonstrating the fix:**
```python
from qiskit import QuantumCircuit, transpile
from qiskit.quantum_info import Statevector
from qiskit.providers.basic_provider import BasicSimulator

qc = QuantumCircuit(3)
qc.h(0)
qc.cx(0, 1)
qc.swap(1, 2)

backend = BasicSimulator()
transpiled1 = transpile(qc, backend, optimization_level=0)
transpiled2 = transpile(qc, backend, optimization_level=2)

# Old method - inconsistent results:
sv1_old = Statevector(transpiled1)
sv2_old = Statevector(transpiled2)
print(sv1_old.equiv(sv2_old))  # False

# New method - consistent results:
sv1_new = Statevector.from_circuit(transpiled1)  
sv2_new = Statevector.from_circuit(transpiled2)
print(sv1_new.equiv(sv2_new))  # True

debasmita2102 committed:
Add Statevector.from_circuit method

Fixes #13621

debasmita2102 committed:
Update release notes with feature description

debasmita2102 committed:
Fix typo in test comments and remove unused import comment

mtreinish commented on releasenotes/notes/add-statevector-from-circuit-9976552cbda50423.yaml:
```diff
@@ -0,0 +1,24 @@
+      
+    An optional ``input_state`` parameter allows starting from a custom
+    initial state instead of the default |00...0⟩ state.
+git add releasenotes/notes/add-statevector-from-circuit-9976552cbda50423.yaml
```
```suggestion
```

debasmita2102 replied:
Removed the typo

mtreinish commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,30 @@ def __init__(
+
+
+        # Get the operator representation of the circuit
+        op = Operator.from_circuit(circuit)
```
This will be problematic because this will limit the width of the circuit you can compute the statevector for. It would be better to iterate over the circuit like the `init` path does but applying the layout transformation with the `qargs` argument directly. You can look at how `Operator.from_circuit` deals with the layout transformations and mirror that logic here for transforming the statevector directly.

debasmita2102 replied:
Done! I've changed the implementation to use `from_instruction `instead of` Operator.from_circuit `to avoid the exponential memory scaling issue.

For the layout handling, I implemented Jake's suggestion using NumPy's tensor reordering (reshape/transpose/reshape) to apply the layout permutation directly on the statevector array. This avoids building the full operator matrix while still correctly handling layout transformations.

Updated in commits f3f30501a (using `from_instruction`) and 4901b8470 (layout handling with NumPy tensor reordering).

mtreinish commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,30 @@ def __init__(
+        Args:
+            circuit (QuantumCircuit): A quantum circuit
+            input_state (Statevector, optional): Input statevector. If None,
+                defaults to the zero state.
```
I think we should probably remove this argument to keep this interface smaller. It's pretty straightforward if the end user wants to do this that they can simply call `input.evolve(qc)` or `input.evolve(Statevector.from_circuit(qc))` themselves. Adding extra arguments here will just increase the complexity and maintenance surface longer term so it would be better to just leave this signature simpler.

debasmita2102 replied:
Done! I've removed the `input_state` parameter to keep the interface simpler as you suggested.

Updated in commit f3f30501a.

mtreinish commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,30 @@ def __init__(
         super().__init__(op_shape=OpShape.auto(shape=shape, dims_l=dims, num_qubits_r=0))
 
+    @classmethod
+    def from_circuit(cls, circuit, input_state=None):
```
Did you want to mirror the options from `Operator.from_layout` for ignoring the layout?

I don't think we want the `initial_layout` or `final_layout` options as personally I think those were mistakes I made in that interface, but the ignore option seems like it would be useful here.

debasmita2102 replied:
Done. I've added the `ignore_set_layout` parameter as you suggested.

The implementation now:

Has `ignore_set_layout=False` parameter (defaults to respecting layout)

Skips the `initial_layout `and `final_layout parameters `

Mirrors the behavior of `Operator.from_circuit` for the layout handling

Updated in commit 5e3ab4f67.

mtreinish reviewed (commented):
Thanks for starting this besides the failing lint and docs build I left a few comments inline.

debasmita2102 committed:
Address review comments: remove input_state, use from_instruction, add ignore_set_layout

- Removed input_state parameter to simplify interface (per review)
- Changed implementation to use from_instruction instead of Operator.from_circuit to avoid exponential memory scaling (per review)
- Added ignore_set_layout parameter to mirror Operator.from_circuit (per review)
- Added layout extraction logic following Operator.from_circuit pattern
- Layout permutation application needs guidance - structure in place
- Updated tests: removed input_state test, added ignore_layout test
- Temporarily skipped transpilation test pending layout implementation guidance
- Applied black formatting
- Updated docstring with examples

debasmita2102 commented:
Hi @mtreinish,

Thanks for the detailed feedback. I've made the changes you suggested-

1. Removed the` input_state` parameter to keep things simpler
2. Switched from `Operator.from_circuit` to `from_instruction` so it doesn't blow up memory on large circuits
3 Added the` ignore_set_layout` parameter like `Operator.from_circuit` has

I'm stuck on the layout permutation part though. I followed how `Operator.from_circuit` extracts the layouts, but I'm not sure how to actually apply those permutations to the statevector without building the full operator matrix (which defeats the purpose).

`Operator.from_circuit ` uses `apply_permutation()` on the matrix, but for a statevector I don't know how to:

- Transform the statevector array with the permutation directly

- Apply the initial_permutation → final_permutation_inverse → initial_permutation_inverse sequence

I tested with the original reproduction code from the issue and it returns False, confirming the layout handling is the missing piece. I've skipped the transpilation consistency test for now. Could you point me in the right direction for handling these permutations on statevector data?

mtreinish commented:
> Operator.from_circuit uses apply_permutation() on the matrix, but for a statevector I don't know how to:
>
> - Transform the statevector array with the permutation directly
> - Apply the initial_permutation → final_permutation_inverse → initial_permutation_inverse sequence
>
> I tested with the original reproduction code from the issue and it returns False, confirming the layout handling is the missing piece. I've skipped the transpilation consistency test for now. Could you point me in the right direction for handling these permutations on statevector data?

The way I would probably go about applying the permutations to the statevector directly is to figure out the swap network needed from the permutation array to get the sequence of swaps to make to apply the permutation. Then apply the appropriate array swap operations on the statevector to apply the swap gate. We don't need to worry about the swap overheads here so you can just use `_synth_permutation_basic` from the the synthesis library to get a circuit of swaps that you can iterate over and look at the qargs for each instruction to get the sequence of qubit swaps to make. You could skip doing the swaps as array operations and just compose the sequence of swap gates from the permutation synthesis output but this would be inefficient.

jakelishman commented:
Matt: we talked about this PR in office hours today.  You don't need to synthesis a swap network to apply a permutation matrix-free, but the principle is sound: a permutation is simply a bunch of row swaps in this case.  It's actually a lot easier than the full operator form, because the operator has an input and output space that may need permuting differently, whereas the statevector only needs to be concerned with the final output.  You can do it either by direct reindexing of the array yourself, or alternatively you can use Numpy's generalised tensor reordering to do it, something like
```python
phys = Statevector(qc)
virt = phys.reshape((2,) * qc.num_qubits).transpose(*qc.layout.final_index_layout()).reshape(-1)
```
(I might have the permutation convenient in `transpose` backwards - it might need inverting.)

That same `transpose` operator is just a bunch of row swaps, though, so we could also directly construct the row reordering and avoid triggering the generalised tensor handling.

debasmita2102 committed:
Implement layout handling for Statevector.from_circuit using NumPy tensor reordering

debasmita2102 commented:
I've implemented the layout handling as suggested  (NumPy tensor reordering approach)

Let me know if you need any adjustments!

debasmita2102 committed:
Update release notes: remove input_state, add ignore_set_layout parameter

debasmita2102 committed:
Remove debug print statements from tests

debasmita2102 committed:
Apply black formatting to test_statevector.py

debasmita2102 committed:
Fix: Remove callable() check and parentheses from final_index_layout

debasmita2102 committed:
Fix: Add pylint disable for false positive on final_index_layout() call

jakelishman commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,75 @@ def __init__(
+        # Handle layout extraction
+        layout = None
+        if not ignore_set_layout:
+            layout = getattr(circuit, "_layout", None)
```
`QuantumCircuit._layout` always exists, so we wouldn't need `getattr`.  But also, we can use the public attribute here, which would be
```python
if not ignore_set_layout:
    layout = circuit.layout
```

jakelishman commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,75 @@ def __init__(
+            final_index_layout = getattr(layout, "final_index_layout", None)
+            if final_index_layout is not None and callable(final_index_layout):
+                # pylint: disable-next=not-callable
+                layout_indices = final_index_layout()
```
Pylint's a bit wrong, but the code is also written in a way that confuses it.

`TranspileLayout.final_index_layout` is a regular method.  We use `getattr`, it's _usually_ for getting instance members, rather than methods, and this is probably what's confusing pylint.  But here, since `layout` is not None then it must be a `TranspileLayout`, so we can just do
```python
layout_indices = layout.final_index_layout()
```

jakelishman commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,75 @@ def __init__(
+            if final_index_layout is not None and callable(final_index_layout):
+                # pylint: disable-next=not-callable
+                layout_indices = final_index_layout()
+                # Reverse for Qiskit's little-endian qubit ordering
```
The convention is actually called "Lsb0" for "least-significant bit is called 0".  Endianness refers to a higher-level concept to do with the ordering of _bytes_ in memory, and doesn't apply to bitstrings.

debasmita2102 replied:
Updated the comment to use the correct terminology "LSb0 (least-significant-bit is 0)" instead of "little-endian".

debasmita2102 committed:
fix(statevector): use final_index_layout for correct qubit order

debasmita2102 commented:
@jakelishman I addressed all previous comments regarding circuit.layout, lsb0, and the direct call to final_index_layout()

debasmita2102 committed:
Merge branch 'main' into feature/add-statevector-from-circuit-clean

Cryoris commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,72 @@ def __init__(
         super().__init__(op_shape=OpShape.auto(shape=shape, dims_l=dims, num_qubits_r=0))
 
+    @classmethod
+    def from_circuit(cls, circuit, ignore_set_layout=False):
```
Could you add inline type hints here and remove them from the docstring? The suggestion here is probably too long for the line to apply it directly
```suggestion
    def from_circuit(cls, circuit: QuantumCircuit, ignore_set_layout: bool = False) -> Statevector:
```

Cryoris commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,72 @@ def __init__(
+                has a layout set, it will be ignored. Defaults to ``False``.
+
+        Returns:
+            Statevector: The statevector representing the circuit evolution
```
```suggestion
            The statevector obtained by applying the circuit on the all-zero
            input state.
```

Cryoris commented on qiskit/quantum_info/states/statevector.py:
```diff
@@ -113,6 +113,72 @@ def __init__(
+        Example:
+            Create a statevector from a transpiled circuit:
+
+            .. code-block:: python
```
We could use 
```suggestion    
            .. plot::
               :include-source:
               :nofigs:
```
to run and test the snippet

Cryoris commented on test/python/quantum_info/states/test_statevector.py:
```diff
@@ -41,6 +41,61 @@
+        self.assertTrue(sv.equiv(expected))
+
+    def test_from_circuit_transpilation_consistency(self):
+        """Test transpilation consistency - the main GitHub issue."""
```
Since this is not a bugfix we don't have to link the issue here. If people are interested they can find the original issue via git blame to get to this PR and then to the issue 🙂

Cryoris commented on test/python/quantum_info/states/test_statevector.py:
```diff
@@ -41,6 +41,61 @@
+        sv1 = Statevector.from_circuit(transpiled_opt0)
+        sv2 = Statevector.from_circuit(transpiled_opt2)
+
+        self.assertTrue(sv1.equiv(sv2))
```
What's not tested yet is that `ignore_set_layout=True` actually does ignore the layout _if one exists_. Could you add this case to check here? E.g.
```suggestion
        self.assertTrue(sv1.equiv(sv2))
        
        sv2_nolayout = Statevector.from_circuit(transpiled_opt2, ignore_set_layout=True)
        self.assertFalse(sv1.equiv(sv2_nolayout))
```

debasmita2102 committed:
Update qiskit/quantum_info/states/statevector.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

debasmita2102 committed:
Update qiskit/quantum_info/states/statevector.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

debasmita2102 committed:
Update qiskit/quantum_info/states/statevector.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

debasmita2102 committed:
Update test/python/quantum_info/states/test_statevector.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

debasmita2102 committed:
Update Statevector.from_circuit transpilation test docstring

debasmita2102 commented:
Thanks @Cryoris  for the review and suggestions, i've resolved them.

Cryoris reviewed (approved):
LGTM thanks for the addition!
